### PR TITLE
Implement Multiply intrinsics of AVX/AVX2/SSE4.1

### PR DESCRIPTION
**Pull request by fiigii**
This PR implements `Multiply` intrinsics of AVX/AVX2/SSE4.1.

### CODE REVIEW

**Comment by fiigii** [comment]
This is a good example to get familiar with simple HW intrinsic implementation where **simple** means single instruction and no memory access.

File: src/jit/emitxarch.cpp [hunk 1]
```diff
@@ -88,14 +88,14 @@ bool emitter::IsDstDstSrcAVXInstruction(instruction ins)
...
-            ins == INS_punpcklwd || ins == INS_punpckhdq || ins == INS_packssdw || ins == INS_packsswb ||
-            ins == INS_packuswb || ins == INS_packusdw || ins == INS_vperm2i128);
+            ins == INS_psubw || ins == INS_psubd || ins == INS_psubq || ins == INS_pmuludq || ins == INS_pmuldq ||
+            ins == INS_pxor || ins == INS_insertps || ins == INS_vinsertf128 || ins == INS_punpckldq ||
+            ins == INS_phaddd || ins == INS_pminub || ins == INS_pminsw || ins == INS_pminsb || ins == INS_pminsd ||
+            ins == INS_pminuw || ins == INS_pminud || ins == INS_pmaxub || ins == INS_pmaxsw || ins == INS_pmaxsb ||
+            ins == INS_pmaxsd || ins == INS_pmaxuw || ins == INS_pmaxud || ins == INS_vinserti128 ||
+            ins == INS_punpckhbw || ins == INS_punpcklbw || ins == INS_punpckhqdq || ins == INS_punpcklqdq ||
+            ins == INS_punpckhwd || ins == INS_punpcklwd || ins == INS_punpckhdq || ins == INS_packssdw ||
+            ins == INS_packsswb || ins == INS_packuswb || ins == INS_packusdw || ins == INS_vperm2i128);
```
Comment (fiigii) [inline src/jit/emitxarch.cpp]: This diff is not clear, I just added `ins == INS_pmuldq` here.

File: src/jit/hwintrinsiccodegenxarch.cpp [hunk 2]
```diff
@@ -318,6 +365,28 @@ void CodeGen::genAVX2Intrinsic(GenTreeHWIntrinsic* node)
...
         }
+
+        case NI_AVX2_Multiply:
+        {
+            op2Reg = op2->gtRegNum;
+            instruction ins;
+            switch (baseType)
+            {
+                case TYP_LONG:
+                    ins = INS_pmuldq;
```
Comment (mikedn) [inline src/jit/hwintrinsiccodegenxarch.cpp]: It looks to me that we should make all this table driven, otherwise we'll end up with a mountain of similar looking code.

File: src/jit/hwintrinsicxarch.cpp [hunk 3]
```diff
@@ -466,6 +479,14 @@ GenTree* Compiler::impAVXIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HAND
             retNode  = gtNewSimdHWIntrinsicNode(TYP_SIMD32, op1, op2, NI_AVX_Add, baseType, 32);
             break;
 
+        case NI_AVX_Multiply:
+            assert(sig->numArgs == 2);
+            op2      = impSIMDPopStack(TYP_SIMD32);
+            op1      = impSIMDPopStack(TYP_SIMD32);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            retNode  = gtNewSimdHWIntrinsicNode(TYP_SIMD32, op1, op2, NI_AVX_Multiply, baseType, 32);
```
Comment (mikedn) [inline src/jit/hwintrinsicxarch.cpp]: You should pass `intrinsic` instead of `NI_AVX_Multiply` to `gtNewSimdHWIntrinsicNode`, then the code for `Add` and `Multiply` will be identical..

File: src/jit/hwintrinsiccodegenxarch.cpp (hunk 2 above)
Comment (fiigii) [reply src/jit/hwintrinsiccodegenxarch.cpp]: I just investigated class `Avx` that has about 75 method names (rather than overloads). Among these methods, about half (~35) are "simple" intrinsics (vector-only parameters) that can be easily table driven. `Avx2` is worse.
Actually, I do not oppose the table driven approach. I am still studying and exploring the implementation, so I am not sure what the code finally look like. For example, I am working on `Load`/`Store` recently. I found that CoreCLR/RyuJIT has had many code to deal with indirect memory access (e.g., `GT_IND`), reusing the code to matching SIMD HW intrinsics is non-trivial (Could you give some suggestions, by the way?).
Further more, high CQ of hardware intrinsic is not trivial neither. I am afraid that the table driven codgen may harm the subsequent optimization work.
In summary, I prefer to implement all (or most) intrinsics with the dirty and quick way at first. Then we can try to refactor it sometime.

Comment (mikedn) [reply src/jit/hwintrinsiccodegenxarch.cpp]: > I found that CoreCLR/RyuJIT has had many code to deal with indirect memory access (e.g., GT_IND), reusing the code to match SIMD HW intrinsics is non-trivial (Could you give some suggestions, by the way?).

Could you explain a bit more what the problem is? I know there are some complications in this area but I'm not sure what exactly are you referring to.

> Further more, high CQ of hardware intrinsic is not trivial neither.

Some examples would be useful. Right now the only obvious "problem" is how to map from `(intrinsicId, baseType)` to `instruction`.

To be clear - I do expect that there will be some special cases that might make the table driven approach problematic at times. But the fact that even something as basic as add/multiply require duplicated code is a bit concerning.

Comment (fiigii) [reply src/jit/hwintrinsiccodegenxarch.cpp]: > Could you explain a bit more what the problem is? I know there are some complications in this area but I'm not sure what exactly you are referring to.

First of all, what should the `Load`/`Store` IR look like? If we use `gtNewSimdHWIntrinsicNode(TYP_SIMD32, pointer, NI_AVX_Load, baseType, 32)`, is it possible to reuse the current indir code?

Comment (mikedn) [reply src/jit/hwintrinsiccodegenxarch.cpp]: I don't think it's feasible for load/stores to be intrinsics, they should be normal indirs having the appropriate type. Attempting to introduce a node that acts like and indir but it is not actually an indir sounds like a lot of trouble.

Also, it's not clear what purpose these load/store intrinsic serve. I expect code like `Vector128<int> v = Unsafe.Read<Vector128<int>>(p);` to work and generate an appropriate load instruction even if no intrinsic is used.

Only the aligned versions of the load/store intrinsics seems to provide value. And we need to figure out how to express those in IR. Probably we'll need to use a GTF flag to indicate that the address is aligned. Or reuse the existing `GTF_IND_UNALIGNED`.

cc @CarolEidt

Comment (fiigii) [reply src/jit/hwintrinsiccodegenxarch.cpp]: > Attempting to introduce a node that acts like and indir but it is not actually an indir sounds like a lot of trouble.
> Only the aligned versions of the load/store intrinsics seems to provide value. 

This concern is not only for `Load`/`Store`/`Aligned`/`Unaligned`, but we also have many other memory access intrinsics, e.g., 
```csharp
public static unsafe void ExtractVector128(byte* address, Vector256<byte> value, byte index);
public static unsafe void MaskStore(double* address, Vector256<double> mask, Vector256<ulong> source);
public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale);
public static unsafe Vector128<short> LoadAlignedNonTemporal(short* address);
```

Comment (mikedn) [reply src/jit/hwintrinsiccodegenxarch.cpp]: Hmm, let's see:
* `LoadAlignedNonTemporal` - this looks like an indir with a `GTF_NONTEMPORAL` flag. Hopefully there's room for a new flag.
* `ExtractVector128` - this seems to be `VEXTRACTI128` and this instruction can store to a register. This looks like it can be modeled as store indir + intrinsic. Though to be able to generate a single instruction the store indir operand (the intrinsic) would need to be contained and I'm not sure if that's possible right now. In the worst case we'll end up with two instructions, it's not that bad.

The other 2 are interesting, they always require memory operands. I suppose it's possible to use the same approach used for `ExtractVector128` with one condition - the JIT optimizer should never attempt to somehow eliminate the indir that the importer/morph creates. Most likely this means that these indir nodes needs to be excluded from CSE and if a local variable is behind the indir then the local variable must be made not enregistrable. And for MaskStore we need to deal with the containment issue.

There's sure going to be a bit of trouble to support these intrinsics by piggy backing on the existing indir nodes but it's likely to be simpler than attempting to hide indirs in intrinsic nodes.
Interlocked nodes already pull this stunt and that causes some problems, see #14557.

File: tests/src/JIT/HardwareIntrinsics/Multiply.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (4creators) [inline tests/src/JIT/HardwareIntrinsics/Multiply.cs]: Where should I place tests in my PRs - per your discussion with @tannergooding they should go to `x86` directory. Is this still holding?

Comment (4creators) [reply tests/src/JIT/HardwareIntrinsics/Multiply.cs]: cc @fiigii @tannergooding

**Comment by CarolEidt** [comment]
@mikedn made some excellent comments. To add my two cents:

- I would agree that a table-driven approach is likely to be cleaner and easier to extend, even when there are complications.

- I would also agree that, as much as possible, we should avoid adding new nodes for indirection, as long as the operation is really a simple indirection, or an operation on a memory location that can be modeled as `OP(IND(addr))`. The `GTF_IND_XXX` flag space is fully subscribed, but we should be able to figure out how to add a flag or otherwise express the non-temporal attribute.

Where indirection is required for the operation, I think the right approach is to define the intrinsic as per its behavior, and then import it as a `GT_HWIntrinsic` with a `GT_IND` child, and set the `GTF_DONT_CSE` at that time. I don't think handling the containment should be all that difficult, but we'll see I suppose.

The `MaskStore`, and `Gather` variants are somewhat different in that they can't really be modeled as a composite such as `STORE_IND(OP(x),addr)`. I think for those the right approach is to define intrinsics.

**Comment by fiigii** [comment]
@CarolEidt @mikedn Thank you so much for the great suggestions. 
- I will change the CodeGen to a table-driven approach.
- I will start with implementing `Load` by `OP(IND(addr))` in a new PR.

BTW, both the above items will be started after the new year, which may influence @tannergooding and @4creators 's work (sorry).

File: src/jit/hwintrinsiccodegenxarch.cpp (hunk 2 above)
Comment (fiigii) [reply src/jit/hwintrinsiccodegenxarch.cpp]: > ExtractVector128 - this seems to be VEXTRACTI128 and this instruction can store to a register.

No, I intentionally exposed `Insert/ExtractVector128` as the memory-access intrinsic to provide the semantics of 16-byte load/store for AVX/AVX2 programming. 
That is a common pattern. Sometimes 32-byte alignment cannot guarantee (but 16-byte alignment can), so AVX 256-bit memory load/store would cause much more cache line split. Developers can use `Avx.Insert/ExtractVector128` to load/store data 128-bit by 128-bit to get better performance.

**Comment by 4creators** [comment]
@fiigii  That's OK
